pyscf/mcscf/apc.py

```diff
@@ -0,0 +1,410 @@
+        condition1 = (nactel > 0)
+        condition2 = (nactel < 2*len(active_idx))
+        condition3 = (len(active_idx) >= num_os)
+        
+        if (condition1 and condition2 and condition3):
+            return True
+        else:
+            self.log.debug("Active space is not reasonable!")
+            if not condition1:
+                self.log.debug("Condition 1 not met")
+            elif not condition2:
+                self.log.debug("Condition 2 not met")
+            elif not condition3:
+                self.log.debug("Condition 3 not met")
+            return False
+        
+    def kernel(self):
+        log = self.log
+        entropies = self.entropies.copy()
+        occ = self.occ.copy()
+        
+        #Change singly occupied orbitals to have larger entropies so they are selected:
+        os_idx = np.where(occ == 1)[0]
+        entropies[os_idx] = np.max(entropies) + 0.01
+        if len(os_idx) > 0:
+            log.info("Singly occupied orbitals found, setting them to have entropy of max + 0.01...")
+        
+        #Start with all orbitals in active space:
+        active_idx = list(range(len(entropies)))
+        inactive_idx = []
+        secondary_idx = []
+        
+        #Size constraint:
+        if isinstance(self.max_size,tuple):
+            nactel,norbs = self.max_size
+            max_size = self._ncsf(nactel,norbs)
+            as_size = self._calc_ncsf(active_idx)
+        else:
+            max_size = self.max_size
+            as_size = len(active_idx)
+                
+        nactel = int(np.sum(np.array(occ)[active_idx]))
+        nactorbs = len(active_idx)
+        log.debug(f"Initial active space of ({nactel},{nactorbs}) has size {as_size}")
+        log.debug(f"Maximum active space size set to {max_size}")
+        
+        #Drop orbitals until size constraint is satisfied:
+        while as_size > max_size:
+            nactel = int(np.sum(np.array(occ)[active_idx]))
+            nactorbs = len(active_idx)
+            log.debug(f"Active space of ({nactel},{nactorbs}) has size {as_size} larger than {max_size}")
+            log.debug(f"Dropping lowest entropy orbital...")
+
+            #Get active orbital entropies
+            active_entropies = entropies[active_idx]
+            ranked_active_idx = [active_idx[i] for i in np.argsort(active_entropies)]
+
+            #Drop lowest orbital in succession, checking for reasonability:
+            active_space_is_reasonable = False
+            tries = 0
+
+            while not active_space_is_reasonable:
+                try:
+                    dropped_idx = ranked_active_idx[tries] #Move to next possibility
+                    dropped_idx_entropy = np.round(entropies[dropped_idx],4)
+                    dropped_idx_occ = int(occ[dropped_idx])
+                except IndexError:
+                    log.error("Not enough orbitals to choose a reasonable active space!")
+                    assert(active_space_is_reasonable)
```

Comment:
active_space_is_reasonable will never be False when reaching this line. What is the reason to assert here?

Reply:
I think it should always be False here, so it's meant as a way to stop execution. Have replaced with
"raise RuntimeError("Not enough orbitals to choose a reasonable active space!")"